### PR TITLE
feat: multiple characters for rule.py

### LINKED ISSUE #204: [REQUEST] support multiple chars for rule printing
Have you checked the issues for a similar suggestions? Yes

**How would you improve Rich?**

Currently for rule.py it only allows the character to have 1 cell length, it would better if it could allow more, like

```python
>>> console.print(Rule("hello", character="-+"))
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ hello -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

**What problem does it solved for you?**

I was just thinking it would improve it and it looks like an ASCII rule, which is cool (like [this](https://github.com/LuRsT/hr#how-to-use-it) program)

---

If this is accepted, I would like to work on it, because I have already tried it out ”locally” (https://repl.it/@hedythedev/rich#rich/rule.py) and it works well

```python
console.print(Rule(text, character="-/-"))
```

BDB784D3-8BA2-4978-9791-6860EEFE0384

### PR DESCRIPTION
**Pull request by hedythedev**
## Type of changes

- [x] New feature

(Removing these other options so progress bar of pr can show proper progress)

## Checklist

- [x] I've run the latest [black](https://github.com/ambv/black) with default args on new code. 
- [ ] I've updated CHANGELOG.md and CONTRIBUTORS.md where appropriate.
- [x] I've added tests for new code.
- [x] I accept that @willmcgugan may be pedantic in the code review.
---
- [x] Take care of super long strings for characters param
- [x] support cjk characters
- [x] support emoji
- [x] update rule function in console.py

## Description
Resolves #204 

Add support for multiple characters param in printing a horizontal rule, supported emoji sizing and cjk printing. Added tests.

### CODE REVIEW

**Comment by hedythedev** [comment]
I was able to get the cell sizes of emoji, but had some trouble with printing emoji, problem seems to happen around [here](https://github.com/hedythedev/rich/blob/e0066e34c2e68ee6503152ab28edfc194bf4cff6/rich/rule.py#L68-L76)
  I think. Also for Chinese characters, this ends up happening `--- 你好！\n ---\n` note the extra `\n` after title

**Comment by willmcgugan** [comment]
There's not much you can do re overlapping emoji I'm afraid, it depends on the terminal. Try other emoji / terminals to see if there is a difference.

Not sure where your extra new line is coming from.

**Comment by hedythedev** [comment]
I've just tried the emoji rule in my own terminal and there aren't any overlaps anymore. But the last emoji gets broken onto the next line, because of the 2 cell length it has. I don't think it will be much of a problem as long as people don't get a too unexpected result when they try to do a rule with emoji. Because using emoji is a bit uncommon in my opinion. I will investigate in cjk character rules immediatly.

**Comment by willmcgugan** [comment]
Ah, I see why there was an extra newline. The text was longer than the width of the terminal and it wrapped.

have a look at `cells.set_cell_size`. It will set the cell width of a string, and handles the case of halving an emoji/cjk by replacing it with a space.

I do think it's worthwhile to handle these edge cases. If only so I have less issues to respond to.

**Comment by hedythedev** [comment]
I have fixed the extra newline, do you think we should add tests for emoji? Since you said the overlapping can differ due to different terminals I haven’t added the, yet

File: rich/rule.py [hunk 1]
```diff
@@ -21,29 +21,37 @@ def __init__(
...
 
+        if cell_len(characters) == 1:
+            chars_len = get_char_size(characters)
+        else:
+            chars_len = 0
+            for i in list(characters):
+                chars_len += get_char_size(i)
         if not self.title:
-            yield Text(character * width, self.style)
+            yield Text(characters * (width // chars_len), self.style)
```
Comment (willmcgugan) [inline rich/rule.py:54]: This may not extend to the end if characters is more than one character.

For instance if the terminal width is 100, and characters is "abc" the rule will be 99 characters since (100 // 3) is 33.

Suggest calculating the characters by adding one extra repetition then truncating to width.

Something like this:

```python
text = Text(characters * (width // chars_len) + 1, self.style) 
text.truncate(width)
```

File: rich/rule.py [hunk 2]
```diff
@@ -56,12 +64,18 @@ def __rich_console__(
...
+                side = Text(characters * (side_width // (chars_len - 1)))
+            side.truncate(side_width - 1)
+            rule_text.append(str(side) + " ", self.style)
             rule_text.append(title_text)
-            rule_text.append(
-                " " + character * (width - cell_len(rule_text.plain) - 1), self.style
-            )
+            rule_text.append(" " + str(side), self.style)
+            if len(rule_text) < width:
+                rule_text.append(characters[0], self.style)
```
Comment (willmcgugan) [inline rich/rule.py:77]: This could lead to the last char not being in sequence. i.e. if characters is "abc", the sequence may not always end in an "a".

How about building the left and right text separately. That way you can calculate the length of the right rule like this:

```python
right_length = width - cell_len(left.plain) - cell_len(title.plain)
```

So you wouldn't have to compensate for an odd width.

File: rich/rule.py (hunk 1 above)
Comment (hedythedev) [reply rich/rule.py]: Needed extra parenthesis `((width // chars_len) + 1)` done

File: rich/rule.py (hunk 2 above)
Comment (hedythedev) [reply rich/rule.py:73]: Will do this next

File: rich/rule.py [hunk 3]
```diff
@@ -12,7 +12,7 @@ class Rule(JupyterMixin):
     
     Args:
         title (Union[str, Text], optional): Text to render in the rule. Defaults to "".
-        character (str, optional): Character used to draw the line. Defaults to "─".
+        characters (str, optional): Character(s) used to draw the line. Defaults to "─".
```
[the remark this answers is not in the document]
Comment (hedythedev) [reply rich/rule.py:15]: How about keeping both, and point character to characters, then have a deprecation warning which is to be removed on the next major bump

[the remark this answers is not in the document]
Comment (willmcgugan) [reply rich/rule.py:15]: Good idea.

[the remark this answers is not in the document]
Comment (hedythedev) [reply rich/rule.py:15]: How will we implement it though? We can probably use the [warnings](https://docs.python.org/3/library/warnings.html) package, or just put a message in the doc string. Not sure if [this](https://stackoverflow.com/questions/49802412/how-to-implement-deprecation-in-python-with-argument-alias) stack overflow question is helpful

[the remark this answers is not in the document]
Comment (willmcgugan) [reply rich/rule.py:15]: I think add `character` and `characters` to both Rule, and Console.rule. 

A warning is not a bad idea, but it feels too trivial to be a warning. I think maybe document it in the docstring, and when v6.0.0 comes around I'll add that to the release notes.

**Comment by willmcgugan** [comment]
Thanks, I'll have a look at your changes soon. Working on 5.1.0 this weekend.

**Comment by willmcgugan** [comment]
@hedythedev This looks good. All we need is to add 'character' back in to the interface and we should be good to merge.

**Comment by hedythedev** [comment]
The only problem now, is that I did not add tests for line 31 (as the coverage states) because then you will also have to remove the test when the next major bump takes place, so the coverage for rule will be 98%. If you think it’s fine I will add them for both rule and console

**Comment by willmcgugan** [comment]
One option would be to do this:

```python
characters = character or characters
```

Without the branch, coverage won't see that as a miss. But you could argue that's a hack to avoid a test.

tbh I don't think it will make much difference either way. Happy to merge. Thanks for your contribution.

I have branch I've been working on and I should be able to get the new rule in the next minor version, hopefully today or tomorrow.

**Comment by hedythedev** [comment]
Would it be necessary to add me to CONTRIBUTORS.md? I think I forgot to do that

**Comment by willmcgugan** [comment]
Yes please. I’d like to keep track of contributors there.
